File: src/types/image_analysis.ts
/**
 * Mirror of `src-tauri/src/models/image_analysis.rs` (MVP 7).
 * Field names em snake_case por compatibilidade com serde.
 */

export type ImageSourceKind =
  | "photo"
  | "video_frame"
  | "evidence"
  | "local_import";

export interface ImageAnalysis {
  id: string;
  occurrence_id: string;
  title: string;
  source_kind: ImageSourceKind;
  source_id: string | null;
  original_relative_path: string;
  original_hash_sha256: string | null;
  analysis_relative_path: string;
  last_export_relative_path: string | null;
  status: string;
  metadata_json: string;
  created_at: string;
  updated_at: string;
}

export interface ImageExport {
  id: string;
  occurrence_id: string;
  image_analysis_id: string;
  output_relative_path: string;
  sidecar_relative_path: string | null;
  hash_sha256: string | null;
  width: number | null;
  height: number | null;
  format: string;
  created_at: string;
  operation_summary_json: string;
}

export interface ImageOperationLog {
  id: number;
  occurrence_id: string;
  image_analysis_id: string;
  action: string;
  details_json: string;
  created_at: string;
}

export interface CreateImageAnalysisInput {
  title: string;
  source_kind: ImageSourceKind;
  source_id?: string | null;
  original_relative_path: string;
  original_hash_sha256?: string | null;
}

export interface ImportLocalImageInput {
  source_path: string;
  title?: string;
}

export interface BackendAdjustments {
  brightness: number;
  contrast: number;
  gamma: number;
  saturation: number;
  grayscale: boolean;
  invert: boolean;
  /** W14.2 — Matiz (rotação de cor, graus). 0 = neutro. Matriz hueRotate
   * do SVG/CSS (preservando luminância) — replicada no backend. */
  hue?: number;
  /** W14.2 — visibilidade de canal (GIMP-style). false zera o canal de
   * saída. Default true. */
  channel_r?: boolean;
  channel_g?: boolean;
  channel_b?: boolean;
}

export type BackendOperation =
  // Geometric (MVP 7)
  | { kind: "rotate_90_cw" }
  | { kind: "rotate_90_ccw" }
  | { kind: "rotate_180" }
  | { kind: "flip_horizontal" }
  | { kind: "flip_vertical" }
  | { kind: "crop"; x: number; y: number; width: number; height: number }
  | { kind: "resize"; width: number; height: number }
  // G12.1 — Edge detection
  | { kind: "edge_sobel"; strength?: number }
  | { kind: "edge_laplacian"; strength?: number }
  | { kind: "edge_canny"; low_threshold?: number; high_threshold?: number }
  // G12.2 — Blur / denoise
  | { kind: "blur_gaussian"; sigma: number }
  | { kind: "blur_median"; radius: number }
  | { kind: "blur_bilateral"; sigma_space: number; sigma_color: number }
  // G12.3 — Enhancement
  | { kind: "clahe"; tile_size?: number; clip_limit?: number }
  | { kind: "histogram_equalize" }
  | { kind: "auto_levels"; percentile_low?: number; percentile_high?: number }
  | { kind: "white_balance_gray_world" }
  // G12.4 — Morphology
  | { kind: "dilate"; radius?: number }
  | { kind: "erode"; radius?: number }
  | { kind: "open"; radius?: number }
  | { kind: "close"; radius?: number }
  // G12.6 — Perspective
  | {
      kind: "perspective";
      src: [[number, number], [number, number], [number, number], [number, number]];
      dst: [[number, number], [number, number], [number, number], [number, number]];
      output_width: number;
      output_height: number;
    }
  // G12 — Misc
  | { kind: "unsharp_mask"; sigma: number; amount?: number }
  | { kind: "threshold"; value: number }
  | {
      kind: "pixelize";
      x: number;
      y: number;
      width: number;
      height: number;
      block_size: number;
    };

export interface ExportImageInput {
  apply_backend_adjustments?: boolean;
  composed_png_base64?: string | null;
  adjustments?: BackendAdjustments | null;
  operations?: BackendOperation[];
  format?: "png" | "jpg";
  operation_summary_json?: string | null;
}

export interface SaveImageAnalysisInput {
  doc: unknown;
  metadata_json?: string;
  title?: string;
}

export interface ImageAnalysisPayload {
  analysis: ImageAnalysis;
  doc: unknown;
}

export interface ImageMetadata {
  width: number;
  height: number;
  mime_type: string | null;
  format_label: string | null;
  size_bytes: number;
  hash_sha256: string | null;
  exif_json: string | null;
  /** G12.8 — Conjunto completo de hashes para chain of custody. */
  hash_set?: ImageHashSet | null;
}

/** G12.8 — Múltiplos hashes (chain of custody pericial). */
export interface ImageHashSet {
  md5: string;
  sha1: string;
  sha256: string;
  sha3_256: string;
}

/** G12.9 — Histograma com 256 bins por canal + estatísticas. */
export interface ImageHistogram {
  red: number[];
  green: number[];
  blue: number[];
  luminance: number[];
  stats: HistogramStats;
}

export interface HistogramStats {
  mean_r: number;
  mean_g: number;
  mean_b: number;
  mean_lum: number;
  stddev_r: number;
  stddev_g: number;
  stddev_b: number;
  stddev_lum: number;
  min_lum: number;
  max_lum: number;
  total_pixels: number;
}

/** G12 — Preview de uma operação (sem persistir). */
export interface ApplyOperationPreviewInput {
  image_base64: string;
  operation: BackendOperation;
}

export interface ApplyOperationPreviewResult {
  image_base64: string;
  width: number;
  height: number;
}

/** G12 — Aplica pilha de ops na imagem original. */
export interface ApplyOperationStackInput {
  relative_path: string;
  adjustments?: BackendAdjustments | null;
  operations?: BackendOperation[];
  as_jpeg?: boolean;
}

export interface ApplyOperationStackResult {
  image_base64: string;
  mime: string;
  width: number;
  height: number;
}

/** G12.21 — Artifact do relatório HTML gerado. */
export interface ImageAnalysisReportArtifact {
  html: string;
  output_relative_path: string;
}

export interface ImageAssetBytes {
  relative_path: string;
  mime_type: string;
  base64: string;
  size_bytes: number;
}

File: src-tauri/src/models/image_analysis.rs
//! Models for the Image Editor (MVP 7 — Editor de Imagem Pericial).
//!
//! Three persisted entities (tabelas em SQLite, migration 009):
//!   - `ImageAnalysis`        — uma sessão `.sicroimage`;
//!   - `ImageExport`          — uma imagem derivada + sidecar JSON;
//!   - `ImageOperationLog`    — log textual (audit) das operações.
//!
//! Plus one transient input/output:
//!   - `CreateImageAnalysisInput` (criação a partir de evidência ou
//!     arquivo local), `ExportImageInput` (parâmetros do export).
//!
//! Wire format é serde-default snake_case — os mirrors TypeScript em
//! `src/types/image_analysis.ts` mantêm os mesmos nomes.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// ImageSourceKind

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ImageSourceKind {
    Photo,
    VideoFrame,
    Evidence,
    LocalImport,
}

impl ImageSourceKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Photo => "photo",
            Self::VideoFrame => "video_frame",
            Self::Evidence => "evidence",
            Self::LocalImport => "local_import",
        }
    }
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "photo" => Some(Self::Photo),
            "video_frame" => Some(Self::VideoFrame),
            "evidence" => Some(Self::Evidence),
            "local_import" => Some(Self::LocalImport),
            _ => None,
        }
    }
}

// ---------------------------------------------------------------------------
// ImageAnalysis (row em `image_analyses`)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageAnalysis {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub title: String,
    pub source_kind: ImageSourceKind,
    pub source_id: Option<String>,
    pub original_relative_path: String,
    pub original_hash_sha256: Option<String>,
    pub analysis_relative_path: String,
    pub last_export_relative_path: Option<String>,
    pub status: String,
    /// JSON object preserved verbatim — pode carregar dimensions,
    /// mime_type, EXIF resumido, etc.
    pub metadata_json: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// ImageExport (row em `image_exports`)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageExport {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub image_analysis_id: Uuid,
    pub output_relative_path: String,
    pub sidecar_relative_path: Option<String>,
    pub hash_sha256: Option<String>,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub format: String,
    pub created_at: DateTime<Utc>,
    pub operation_summary_json: String,
}

// ---------------------------------------------------------------------------
// ImageOperationLog (row em `image_operation_logs`)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageOperationLog {
    pub id: i64,
    pub occurrence_id: Uuid,
    pub image_analysis_id: Uuid,
    pub action: String,
    pub details_json: String,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// Inputs (Tauri commands)

#[derive(Debug, Clone, Deserialize)]
pub struct CreateImageAnalysisInput {
    pub title: String,
    pub source_kind: ImageSourceKind,
    pub source_id: Option<String>,
    /// Caminho relativo ao workspace. O backend valida com
    /// `sanitize_relative_path`.
    pub original_relative_path: String,
    /// Hash conhecido da imagem original (opcional). Quando ausente,
    /// o backend tenta computar a partir do arquivo no workspace.
    pub original_hash_sha256: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ImportLocalImageInput {
    /// Caminho absoluto da imagem escolhida pelo usuário no diálogo do SO.
    /// O backend copia para `imagens/originais/` e cria a análise.
    pub source_path: String,
    pub title: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct ExportImageInput {
    /// Quando `true` o pipeline tenta aplicar os ajustes no Rust
    /// (brilho/contraste/grayscale/inverter/rotação/flip/crop). Quando
    /// `false`, exporta o PNG já composto enviado pelo frontend (após
    /// aplicar anotações via Konva).
    #[serde(default)]
    pub apply_backend_adjustments: bool,
    /// PNG já composto pelo frontend (base64, opcional). Quando
    /// presente, o backend usa-o como bytes finais; o original e os
    /// ajustes ficam só na sessão.
    #[serde(default)]
    pub composed_png_base64: Option<String>,
    /// Ajustes que o backend deve aplicar (quando
    /// `apply_backend_adjustments=true`). O frontend é a fonte de
    /// verdade da sessão; o backend só re-aplica para o derivado.
    #[serde(default)]
    pub adjustments: Option<BackendAdjustments>,
    /// Lista de operações geométricas a aplicar (na ordem).
    #[serde(default)]
    pub operations: Vec<BackendOperation>,
    /// "png" (padrão) | "jpg".
    #[serde(default)]
    pub format: Option<String>,
    /// JSON livre — vai para o sidecar como "summary".
    #[serde(default)]
    pub operation_summary_json: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct BackendAdjustments {
    /// Brilho em [-100, 100]. 0 = neutro.
    #[serde(default)]
    pub brightness: f32,
    /// Contraste em [-100, 100]. 0 = neutro.
    #[serde(default)]
    pub contrast: f32,
    /// Gamma. 1.0 = neutro. Aplicado como `pixel ^ (1/gamma)`.
    #[serde(default = "default_gamma")]
    pub gamma: f32,
    /// Saturação em [-100, 100]. 0 = neutro.
    #[serde(default)]
    pub saturation: f32,
    #[serde(default)]
    pub grayscale: bool,
    #[serde(default)]
    pub invert: bool,
}

impl Default for BackendAdjustments {
    fn default() -> Self {
        Self {
            brightness: 0.0,
            contrast: 0.0,
            gamma: 1.0,
            saturation: 0.0,
            grayscale: false,
            invert: false,
        }
    }
}

fn default_gamma() -> f32 {
    1.0
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum BackendOperation {
    // ---------- Geometric (MVP 7 — original)
    Rotate90Cw,
    Rotate90Ccw,
    Rotate180,
    FlipHorizontal,
    FlipVertical,
    Crop {
        x: u32,
        y: u32,
        width: u32,
        height: u32,
    },
    Resize {
        width: u32,
        height: u32,
    },

    // ---------- G12.1 — Edge detection
    /// Detecção de bordas Sobel (gradiente).
    EdgeSobel {
        /// Intensidade da resposta em [0.0, 4.0]. 1.0 = neutro.
        #[serde(default = "default_one")]
        strength: f32,
    },
    /// Detecção de bordas Laplaciano (5x5 kernel).
    EdgeLaplacian {
        #[serde(default = "default_one")]
        strength: f32,
    },
    /// Canny (binário). Limites configuráveis.
    EdgeCanny {
        #[serde(default = "default_low_threshold")]
        low_threshold: f32,
        #[serde(default = "default_high_threshold")]
        high_threshold: f32,
    },

    // ---------- G12.2 — Blur / denoise
    /// Gaussian blur. `sigma` controla o spread; raio = ceil(3*sigma).
    BlurGaussian {
        sigma: f32,
    },
    /// Median filter — remove outliers preservando bordas. `radius` em pixels.
    BlurMedian {
        radius: u32,
    },
    /// Bilateral (simplificado) — suaviza preservando edges. Custoso.
    BlurBilateral {
        sigma_space: f32,
        sigma_color: f32,
    },

    // ---------- G12.3 — Enhancement
    /// CLAHE — equalização adaptativa por blocos (tiles).
    Clahe {
        #[serde(default = "default_tile_size")]
        tile_size: u32,
        #[serde(default = "default_clip_limit")]
        clip_limit: f32,
    },
    /// Histogram equalization global na luminância.
    HistogramEqualize,
    /// Auto-levels — estica histograma por canal (RGB) para [percentile_low, percentile_high].
    AutoLevels {
        #[serde(default = "default_percentile_low")]
        percentile_low: f32,
        #[serde(default = "default_percentile_high")]
        percentile_high: f32,
    },
    /// White balance gray-world.
    WhiteBalanceGrayWorld,

    // ---------- G12.4 — Morphology (em luminância)
    /// Dilatação morfológica (kernel quadrado 3x3 ou 5x5).
    Dilate {
        #[serde(default = "default_radius")]
        radius: u32,
    },
    /// Erosão morfológica.
    Erode {
        #[serde(default = "default_radius")]
        radius: u32,
    },
    /// Abertura morfológica = erode → dilate.
    Open {
        #[serde(default = "default_radius")]
        radius: u32,
    },
    /// Fechamento morfológico = dilate → erode.
    Close {
        #[serde(default = "default_radius")]
        radius: u32,
    },

    // ---------- G12.6 — Perspective
    /// Correção de perspectiva — 4 cantos source → 4 cantos destination.
    /// Coordenadas em pixels da imagem original. Output: `output_width` x `output_height`.
    Perspective {
        src: [[f32; 2]; 4],
        dst: [[f32; 2]; 4],
        output_width: u32,
        output_height: u32,
    },

    // ---------- G12 — Extras úteis
    /// Unsharp mask — sharpening clássico.
    UnsharpMask {
        sigma: f32,
        #[serde(default = "default_one")]
        amount: f32,
    },
    /// Threshold simples (binarização).
    Threshold {
        value: u8,
    },
    /// Pixelize uma região (anonimização).
    Pixelize {
        x: u32,
        y: u32,
        width: u32,
        height: u32,
        block_size: u32,
    },
}

fn default_one() -> f32 {
    1.0
}
fn default_low_threshold() -> f32 {
    50.0
}
fn default_high_threshold() -> f32 {
    150.0
}
fn default_tile_size() -> u32 {
    8
}
fn default_clip_limit() -> f32 {
    2.0
}
fn default_percentile_low() -> f32 {
    1.0
}
fn default_percentile_high() -> f32 {
    99.0
}
fn default_radius() -> u32 {
    1
}

// ---------------------------------------------------------------------------
// Image metadata (returned by `get_image_metadata`)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageMetadata {
    pub width: u32,
    pub height: u32,
    pub mime_type: Option<String>,
    pub format_label: Option<String>,
    pub size_bytes: u64,
    pub hash_sha256: Option<String>,
    /// JSON serialised — vazio quando não houver EXIF lido.
    pub exif_json: Option<String>,
    /// G12.8 — Conjunto completo de hashes pericial (opcional).
    /// Computado só quando o caller pede `compute_hash=true` e o backend
    /// suporta — fica `None` em metadados leves.
    #[serde(default)]
    pub hash_set: Option<HashSet>,
}

/// G12.8 — Conjunto de hashes para chain of custody pericial.
///
/// MD5 é matematicamente comprometido, mas ainda é exigido por convenção
/// em muitos laudos institucionais. SHA-1 idem. SHA-256 é o atual padrão
/// recomendado. SHA-3-256 é a próxima geração (Keccak), oferecido como
/// "future-proofing" para laudos que precisem sobreviver décadas.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HashSet {
    pub md5: String,
    pub sha1: String,
    pub sha256: String,
    pub sha3_256: String,
}

/// G12.9 — Histograma + estatísticas básicas de uma imagem.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageHistogram {
    /// 256 bins (count de pixels com cada valor 0..255), canal R.
    pub red: Vec<u32>,
    pub green: Vec<u32>,
    pub blue: Vec<u32>,
    /// Luminância calculada via 0.299R + 0.587G + 0.114B.
    pub luminance: Vec<u32>,
    pub stats: HistogramStats,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistogramStats {
    /// Per-channel mean (0..255).
    pub mean_r: f32,
    pub mean_g: f32,
    pub mean_b: f32,
    pub mean_lum: f32,
    /// Per-channel stddev.
    pub stddev_r: f32,
    pub stddev_g: f32,
    pub stddev_b: f32,
    pub stddev_lum: f32,
    /// Min / Max do canal de luminância (útil para gauge de dinâmica).
    pub min_lum: u8,
    pub max_lum: u8,
    pub total_pixels: u32,
}

// ---------------------------------------------------------------------------
// Asset bytes (returned by `read_image_asset` — base64 igual ao do MVP 4).

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageAssetBytes {
    pub relative_path: String,
    pub mime_type: String,
    pub base64: String,
    pub size_bytes: u64,
}
